network/components/mux_rx.c

```diff
@@ -60,14 +60,28 @@ void rx_return(void)
             assert(!err);
 
             buffer.phys_or_offset = buffer.phys_or_offset - buffer_data_paddr;
-            microkit_arm_vspace_data_invalidate(buffer.phys_or_offset + buffer_data_vaddr, buffer.phys_or_offset + buffer_data_vaddr + buffer.len);
+            const uintptr_t buffer_vaddr = buffer.phys_or_offset + buffer_data_vaddr;
 
-            int client = get_client((struct ethernet_header *) (buffer.phys_or_offset + buffer_data_vaddr));
+            // Cache invalidate after DMA write, so we don't read stale data.
+            // This is needed even if we invalidate before the DMA write as it
+            // could have been speculatively fetched before the DMA write.
+            microkit_arm_vspace_data_invalidate(buffer_vaddr, buffer_vaddr + buffer.len);
+
+            int client = get_client((struct ethernet_header *) buffer_vaddr);
             if (client >= 0) {
                 err = enqueue_used(&state.rx_ring_clients[client], buffer);
                 assert(!err);
                 notify_clients[client] = true;
             } else {
+                // Cache invalidate before DMA write to discard dirty
+                // cachelines, so they don't overwrite received data.
+                //
+                // TODO I think this is technically not needed? as we know we
+                // haven't written to the buffer, and stale caches will get
+                // invalidated by the previous invalidate when we next receive
+                // data to this buffer.
```

Comment:
Highlighing this TODO

Reply:
We should discuss and sort this out rather than putting a TODO.

Reply:
Had a discussion with Courtney, planning to remove this invalidate.